### ISSUE TITLE
SSphp selects the first <AuthenticatingAuthority> context when generating EPTID (or others requiring computation)

### ISSUE THREAD

**Issue by spaetow** [simplesamlphp/simplesamlphp#1431]
**Describe the bug**

When a SAML response (and assertion) contains multiple <AuthenticatingAuthority> statements, it appears that SSphp simply selects the first AA. It does not do any matching to possibly the <Issuer> when it attempts to generate the EPTID. 

This appears to be the case when an Azure IdP is put behind a Shibboleth frontend.

**To Reproduce**

Put a Shibboleth frontend in front of an Azure IdP, maybe? I can provide provide a partial assertion from our member who is experiencing this?

**Expected behavior**

I'd expect the correct <AuthenticatingAuthority> to be selected in this order:

1. If there is only one <AuthenticatingAuthority>, use that.
2. If there is more than one <AuthenticatingAuthority>, select the one that matches the <Issuer> in the statement
3. If there is more than one <AuthenticatingAuthority> but there is no match with the <Issuer>, then either:
    a) Select the first <AuthenticatingAuthority>
    b) Select the last <AuthenticatingAuthority>

**Screenshots or logs**

I don't have logs or screenshots at this time.

**Additional context**

None at this time.

**Comment by thijskh** [comment]
Thanks for the thorough report. What method are you using to generate the epTID?

**Comment by jiscfoo** [comment]
> Thanks for the thorough report. What method are you using to generate the epTID?

In the case that @spaetow is referring to, the eptid is coming from a Shibboleth IdP v4 configured to "proxy" the Authentication step off to another SAML IdP (in our case, Azure). When Shib returns the final assertion to the SP, it contains a `AuthenticatingAuthority` element advertising that the proxy has happened. 
```xml
<saml2:AuthnStatement AuthnInstant="..." SessionIndex="...">
    <saml2:SubjectLocality Address="..."/>
    <saml2:AuthnContext>
        <saml2:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</saml2:AuthnContextClassRef>
        <saml2:AuthenticatingAuthority>upstream idp entityid</saml2:AuthenticatingAuthority>
    </saml2:AuthnContext>
</saml2:AuthnStatement>
```

The `Issuer` remains unchanged:

```xml
<saml2:Issuer xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">proxying idp entityid</saml2:Issuer>
```


I observed (using the authn check) that SimpleSAMLphp generates an array to represent this:

```
    "saml:AuthenticatingAuthority": [
        "upstream idp entityid",
        "proxying idp entityid"
    ],
```

The SmartID module (which I think is being used in this case) then selects the `0` element to use in generating the attribute. This is not configurable and, arguably, wrong -- should this actually be selecting the "last" element of the array (as the "first" IdP up in the chain)?

```
if (($this->add_authority) && (isset($state['saml:AuthenticatingAuthority'][0]))) {
    return ($this->add_candidate ? $idCandidate . ':' : '') . $attributes[$idCandidate][0] . '!' .
        $state['saml:AuthenticatingAuthority'][0];
```

https://github.com/simplesamlphp/simplesamlphp-module-smartattributes/blob/master/lib/Auth/Process/SmartID.php#L115

The result of this is a SmartID of, for example:

```
eduPersonTargetedID:<value>!upstream idp entityid
```

**Comment by tvdijen** [comment]
The SAML specification has left the order of the `AuthenticatingAuthority`-element unspecified..

**Comment by jiscfoo** [comment]
Understood… the SAML specification does not include `Issuer` in `AuthenticatingAuthority` but the current behaviour of squashing them together and making assumptions based on the resulting order seems _wrong_.

If the ordering of `AuthenticatingAuthority` is undefined then the sensible default would be to use the `Issuer` which is not liable to reordering changes and, possibly, allow the SP operator to define some rules on how to interpret the list.

If the unordered list concept is taken one step further then the IdP is _allowed_ to re-order the elements in the `<AuthenticatingAuthority>` list but that would break the SmartID assumption that `[0]` is the "top" IdP.

**Comment by jiscfoo** [comment]
Possibly related to: https://github.com/simplesamlphp/simplesamlphp-module-smartattributes/issues/2

**Comment by c00kiemon5ter** [comment]
Hello everyone,

the SAML core specification defines certain "Proxying Processing Rules" (see section `3.4.1.5.1`). There it is defined that (highlight is mine)

> - The `<saml:AuthnStatement>` in the new assertion MUST include a `<saml:AuthnContext>` element containing a `<saml:AuthenticatingAuthority>` element referencing the identity provider to which the proxying identity provider referred the presenter. **If the original assertion contains `<saml:AuthnContext>` information that includes one or more `<saml:AuthenticatingAuthority>` elements, those elements SHOULD be included in the new assertion, with the new element placed after them.**

so, "the new element" will be the last and this means that SSphp should be looking there instead of the first element in the list. Thus, while ordering is still not defined in absolute terms, in the proxying scenarios we can make that assumption.

I hope this helps move the issue forward.

**Comment by thijskh** [comment]
I've proposed a fix for the smartattributes module.

**Comment by jiscfoo** [comment]
> I've proposed a fix for the smartattributes module.

Thanks for looking at this @thijskh … am I correct in reading the new code that it will, by default, now always select the `Issuer` from the assertion (which has been placed at the end of the `state` array)?

**Comment by thijskh** [comment]
Yes, which seems the sensible approach to me, I don't think you want to use anything other than the last hop?
